pkg/upgrade/job/container.go

```diff
@@ -0,0 +1,102 @@
+package job
+
+import (
+	"path/filepath"
+	"strings"
+
+	upgradeapiv1 "github.com/rancher/system-upgrade-controller/pkg/apis/upgrade.cattle.io/v1"
+	"github.com/rancher/wrangler/pkg/name"
+	corev1 "k8s.io/api/core/v1"
+)
+
+type containerOpt func(*corev1.Container)
+
+func withSecrets(secrets []upgradeapiv1.SecretSpec) containerOpt {
+	return func(container *corev1.Container) {
+		for _, secret := range secrets {
+			secretVolumeName := name.SafeConcatName("secret", secret.Name)
+			secretVolumePath := secret.Path
+			if secretVolumePath == "" {
+				secretVolumePath = filepath.Join("/run/system-upgrade/secrets", secret.Name)
+			} else if secretVolumePath[0:1] != "/" {
+				secretVolumePath = filepath.Join("/run/system-upgrade/secrets", secretVolumePath)
+			}
+			container.VolumeMounts = append(container.VolumeMounts, corev1.VolumeMount{
+				Name:      secretVolumeName,
+				MountPath: secretVolumePath,
+				ReadOnly:  true,
+			})
+		}
+	}
+}
+
+func withSecurityContext(securityContext *corev1.SecurityContext) containerOpt {
+	return func(container *corev1.Container) {
+		container.SecurityContext = securityContext
+	}
+}
+
+func withImageTag(tag string) containerOpt {
+	return func(container *corev1.Container) {
+		image := container.Image
+		if p := strings.Split(image, `:`); len(p) > 1 {
+			image = strings.Join(p[0:len(p)-1], `:`)
```

Comment:
I am wondering if the code also check if the user specified the tag in the spec.image for example:
```
Spec:
  Image: rancher/k3s-upgrade:v1.17.2
```
correct me if i am wrong but according to the code it will ignore the v1.17.2 value if the version is not specified

Reply:
@galal-hussein  yes, this will ignore a tag on the image if supplied for the `upgrade` container because the tag for this image should be the resolved version aka `status.latestVersion`